metaseq/tasks/streaming_language_modeling.py

```diff
@@ -213,26 +216,60 @@ def _alpha_sampling(self, datasets, corpora, epoch=1):
             [len(d) for d in datasets],
             dtype=float,
         )
+        sample_probs = self._get_sample_prob(dataset_lengths)
         logger.info(f"loaded total {dataset_lengths.sum()} blocks for all corpora")
-        if self.args.pretrain_data_sampling_prob >= 0:
-            assert self.args.pretrain_data_sampling_prob < 1.0
-
-            def prefix_match_index(l):
-                prefix = "pretrain__"
-                for i, s in enumerate(l):
-                    if prefix in s:
-                        return i
-                return -1
-
-            pretrain_corpus_index = prefix_match_index(corpora)
-            assert pretrain_corpus_index != -1
-            tmp_dataset_lengths = np.copy(dataset_lengths)
-            tmp_dataset_lengths[pretrain_corpus_index] = 0
-            sample_probs = self._get_sample_prob(tmp_dataset_lengths)
-            sample_probs = sample_probs * (1 - self.args.pretrain_data_sampling_prob)
-            sample_probs[pretrain_corpus_index] = self.args.pretrain_data_sampling_prob
+
+        def compute_benchmark_prob(sample_probs):
+            per_benchmark_prob = {}
+            for i, s in enumerate(corpora):
+                benchmark = s.split('__')[0]
+                if self.args.equalize_cluster_probs:
+                    cluster = s.split('__')[1]
+                    per_benchmark_prob.setdefault(benchmark, {})
+                    per_benchmark_prob[benchmark].setdefault(cluster, 0.0)
+                    per_benchmark_prob[benchmark][cluster] += sample_probs[i]
+                else:
+                    if benchmark not in per_benchmark_prob:
+                        per_benchmark_prob[benchmark] = 0.0
+                    per_benchmark_prob[benchmark] += sample_probs[i]
+            return per_benchmark_prob
+
+        per_benchmark_prob = compute_benchmark_prob(sample_probs)
+        logger.info(f'Initial data proportions: {json.dumps(per_benchmark_prob)}')
+
+        data_sampling_prob = json.loads(self.args.data_sampling_prob)
+
+        # If proportions are not specified for all the benchmarks
+        # scale up/down the remaining proportions uniformly
+        total_benchmarks = len(per_benchmark_prob)
+        total_assigned_prob = sum(data_sampling_prob.values())
+        total_assigned_benchmarks = len(data_sampling_prob)
+
+        # Assign probs for the benchmarks not present in the dictionary
+        for benchmark in per_benchmark_prob:
+            if benchmark not in data_sampling_prob:
+                benchmark_prob = per_benchmark_prob[benchmark] if not self.args.equalize_cluster_probs else sum(per_benchmark_prob[benchmark].values())
+                data_sampling_prob[benchmark] = benchmark_prob * (1.0 - total_assigned_prob)
```

Comment:
Previously with pre-train data proportion, we use to keep the task eps but not balance the benchmarks. But, now it seems like that case is not possible given that we always rebalance the benchmarks when only pre-train proportion is mentioned in data_sampling_prob arg. I think that is fine, but just want to make sure that's the case.